src/main/java/org/jabref/logic/importer/fetcher/ResearchGate.java

```diff
@@ -0,0 +1,268 @@
+
+        // DOI search
+        Optional<String> title = entry.getField(StandardField.TITLE);
+        Optional<DOI> doi = entry.getField(StandardField.DOI).flatMap(DOI::parse);
+
+        // Retrieve PDF link
+        String linkForSearch;
+        Document html;
+        try {
+            if (title.isPresent()) {
+                LOGGER.trace("Search by Title");
+                linkForSearch = getURLByString(title.get());
+                Connection connection = Jsoup.connect(linkForSearch);
+                html = connection
+                        .cookieStore(connection.cookieStore())
+                        .userAgent(URLDownload.USER_AGENT)
+                        .referrer("www.google.com")
+                        .ignoreHttpErrors(true)
+                        .get();
+            } else if (doi.isPresent()) {
+                LOGGER.trace("Search by DOI");
+                // Retrieve PDF link
+                Connection connection = Jsoup.connect(getURLByDoi(doi.get()));
+                html = connection
+                        .cookieStore(connection.cookieStore())
+                        .userAgent(URLDownload.USER_AGENT)
+                        .ignoreHttpErrors(true)
+                        .get();
+            } else {
+                return Optional.empty();
+            }
+        } catch (FetcherException | NullPointerException e) {
+            LOGGER.debug("ResearchGate server is not available");
+            return Optional.empty();
+        }
+        Elements eLink = html.getElementsByTag("section");
+        String link = eLink.select("a[href^=https]").select("a[href$=.pdf]").attr("href");
+        LOGGER.debug("PDF link: {}", link);
+
+        if (link.contains("researchgate.net")) {
+            return Optional.of(new URL(link));
+        }
+        return Optional.empty();
+    }
+
+    String getURLByString(String query) throws IOException, FetcherException, NullPointerException {
+        URIBuilder source;
+        String link;
+        try {
+            source = new URIBuilder(SEARCH);
+            source.addParameter("type", "publication");
+            source.addParameter("query", query);
+
+            URLDownload urlDownload = new URLDownload(source.toString());
+            urlDownload.getCookieFromUrl();
+            Document html = Jsoup.connect(source.toString())
+                                 .userAgent(URLDownload.USER_AGENT)
+                                 .referrer("www.google.com")
+                                 .ignoreHttpErrors(true)
+                                 .get();
+
+            link = HOST + Objects.requireNonNull(html.getElementById("content"))
+                                 .select("a[href^=publication/]")
+                                 .attr("href");
+            if (link.contains("?")) {
+                link = link.substring(0, link.indexOf("?"));
+            }
+        } catch (URISyntaxException e) {
+            return null;
```

Comment:
Not happy about returning null here. Maybe empty string?

Reply:
I refactored the Code to return Optional<String>